pkg/sqlite/graphloader.go

```diff
@@ -0,0 +1,173 @@
+	return &SQLGraphLoader{
+		Querier:     NewSQLLiteQuerierFromDb(db),
+		PackageName: name,
+	}, nil
+}
+
+func (g *SQLGraphLoader) Generate() (*registry.Package, error) {
+	ctx := context.TODO()
+	defaultChannel, err := g.Querier.GetDefaultPackage(ctx, g.PackageName)
+	if err != nil {
+		return nil, err
+	}
+
+	channelEntries, err := g.Querier.GetChannelEntriesFromPackage(ctx, g.PackageName)
+	if err != nil {
+		return nil, err
+	}
+
+	channels, err := g.GraphFromEntries(channelEntries)
+	if err != nil {
+		return nil, err
+	}
+
+	return &registry.Package{
+		Name:           g.PackageName,
+		DefaultChannel: defaultChannel,
+		Channels:       channels,
+	}, nil
+}
+
+// TODO get bundle path and version
+
+// GraphFromEntries builds the graph from a set of channel entries
+func (g *SQLGraphLoader) GraphFromEntries(channelEntries []ChannelEntryNode) ([]registry.Channel, error) {
+	var channels []registry.Channel
+	var channelToBundles = make(map[string][]registry.OperatorBundle)
+
+	for _, entry := range channelEntries {
+		newBundle := registry.OperatorBundle{
+			Version:         entry.Version,
+			CsvName:         entry.BundleName,
+			BundlePath:      entry.BundlePath,
+			ReplacesBundles: []registry.OperatorBundle{},
+			Replaces:        []registry.BundleRef{},
+		}
+
+		replaces := registry.BundleRef{
+			BundlePath: entry.BundlePath,
+			Version:    entry.ReplacesVersion,
+			CsvName:    entry.Replaces,
+		}
+
+		if !replaces.IsEmptyRef() {
+			newBundle.Replaces = append(newBundle.Replaces, replaces)
+		}
+
+		if bundles, ok := channelToBundles[entry.ChannelName]; !ok {
+			channelToBundles[entry.ChannelName] = []registry.OperatorBundle{newBundle}
+		} else {
+			// if newBundle is in the channel then append replaces to that newBundle
+			// else insert newBundle
+			bundle := getBundle(bundles, entry.BundleName)
+			if bundle != nil {
+				bundle.Replaces = append(bundle.Replaces, replaces)
+			} else {
+				channelToBundles[entry.ChannelName] = append(channelToBundles[entry.ChannelName], newBundle)
+			}
+		}
+	}
+
+	// bundleref to operatorbundle
+	for _, bundles := range channelToBundles {
+		for _, bundle := range bundles {
+			for _, ref := range bundle.Replaces {
+				replacesBundle := getBundle(bundles, ref.CsvName)
+				if replacesBundle != nil {
+					bundle.ReplacesBundles = append(bundle.ReplacesBundles, *replacesBundle)
+				}
+			}
+		}
+	}
+
+	for chName, bundles := range channelToBundles {
+		head := getHeadBundleRefForChannel(bundles)
+
+		channel := registry.Channel{
+			Name:            chName,
+			OperatorBundles: bundles,
+			Head:            *head,
+		}
+
+		channels = append(channels, channel)
+	}
+
+	return channels, nil
+}
+
+func getBundle(bundles []registry.OperatorBundle, name string) *registry.OperatorBundle {
+	for _, b := range bundles {
+		if b.CsvName == name {
+			return &b
+		}
+	}
+	return nil
+}
+
+func getHeadBundleRefForChannel(bundles []registry.OperatorBundle) *registry.BundleRef {
```

Comment:
This assumes an already valid graph, which might not be true of the places you're intending to call this

Reply:
This specific implementation is reading from the sql layer. The assumption here is that the graph is valid because we wouldn't persist it in the first place otherwise.

That being said, I can totally as we flesh this out attempting to do some generation validation. Maybe that's what we should tackle next, and when the insert is written we include that as part of this function?